pkg/distributor/distributor.go

```diff
@@ -270,22 +282,33 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
 	}
 	select {
 	case err := <-tracker.err:
-		return nil, err
+		_, cd := util.IsCodedError(err)
+		return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+			Code: int32(cd),
+			Body: []byte(err.Error()),
+		})
 	case <-tracker.done:
-		return &logproto.PushResponse{}, validationErr
+		if errSampler.Size() > 0 {
+			return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+				Code: int32(errSampler.Code()),
+				Body: []byte(errSampler.Error()),
+			})
+		}
+		return &logproto.PushResponse{}, nil
 	case <-ctx.Done():
-		return nil, ctx.Err()
+		return nil, httpgrpc.ErrorFromHTTPResponse(&httpgrpc.HTTPResponse{
+			Code: int32(http.StatusInternalServerError),
+			Body: []byte(ctx.Err().Error()),
```

Comment:
you may have some error in the errorSampler ? may be you could Add to it and then use it ?